# Autogenerated Azimuth analysis script
# Sections related to visualizations are discarded

library(Seurat)
library(Azimuth)
library(dplyr)
library(feather)

start_time <- Sys.time()


data_path <- "/home/rohan/Local/datasets/"

query_data_path <- paste0(data_path, "human_mtg_query/")
query_files <- list.files(query_data_path, pattern = "*.RDS")
df <- read_feather(paste0(data_path, "MTG_AD_metadata_2022-04-13.feather"))

# Load the reference
reference <- LoadReference(path = paste0(data_path, "/human_mtg/ref_proc/"))
count <- 0

for (this_query_file in query_files) {
    count <- count + 1
    print(paste('Starting file #',count,'filename',this_query_file))

    # Load the query object for mapping
    query <- LoadFileInput(path = paste0(query_data_path, this_query_file))
    query <- ConvertGeneNames(
        object = query,
        reference.names = rownames(x = reference$map),
        homolog.table = 'https://seurat.nygenome.org/azimuth/references/homologs.rds'
    )

    # Restricting based on metadata provided by Kyle
    keep_df <- df %>% filter(sample_id %in% colnames(query))
    cells.use <- keep_df$sample_id
    query <- query[, cells.use]

    # Calculate nCount_RNA and nFeature_RNA if the query does not
    # contain them already
    if (!all(c("nCount_RNA", "nFeature_RNA") %in% c(colnames(x = query[[]])))) {
        calcn <- as.data.frame(x = Seurat:::CalcN(object = query))
        colnames(x = calcn) <- paste(colnames(x = calcn), "RNA", sep = '_')
        query <- AddMetaData(object = query, metadata = calcn)
        rm(calcn)
    }

    # Calculate percent mitochondrial genes if the query contains genes
    # matching the regular expression "^MT-"
    if (any(grepl(pattern = '^MT-', x = rownames(x = query)))) {
        query <- PercentageFeatureSet(
            object = query,
            pattern = '^MT-',
            col.name = 'percent.mt',
            assay = "RNA"
        )
    }


    # Preprocess with SCTransform
    query <- SCTransform(
        object = query,
        assay = "RNA",
        new.assay.name = "refAssay",
        residual.features = rownames(x = reference$map),
        reference.SCT.model = reference$map[["refAssay"]]@SCTModel.list$refmodel,
        method = 'glmGamPoi',
        ncells = 2000,
        n_genes = 2000,
        do.correct.umi = FALSE,
        do.scale = FALSE,
        do.center = TRUE
    )

    # Find anchors between query and reference
    anchors <- FindTransferAnchors(
        reference = reference$map,
        query = query,
        k.filter = NA,
        reference.neighbors = "refdr.annoy.neighbors",
        reference.assay = "refAssay",
        query.assay = "refAssay",
        reference.reduction = "refDR",
        normalization.method = "SCT",
        features = intersect(rownames(x = reference$map),
            VariableFeatures(object = query)),
        dims = 1:50,
        n.trees = 20,
        mapping.score.k = 100
    )

    # Transfer cell type labels
    # Transferred labels are in metadata columns named "predicted.*"
    # The maximum prediction score is in a metadata column named "predicted.*.score"
    # The prediction scores for each class are in an assay named "prediction.score.*"
    refdata <- lapply(X = c("cluster", "subclass"), function(x) {
        reference$map[[x, drop = TRUE]]
    })
    names(x = refdata) <- c("cluster", "subclass")
    if (FALSE) {
        refdata[["impADT"]] <- GetAssayData(object = reference$map[['ADT']],
                                            slot = 'data')
    }
    query <- TransferData(
        reference = reference$map,
        query = query,
        dims = 1:50,
        anchorset = anchors,
        refdata = refdata,
        n.trees = 20,
        store.weights = TRUE
    )

    res_df <-
        query@meta.data[c(
            "sample_id",
            "predicted.cluster",
            "predicted.subclass",
            "predicted.subclass.score",
            "predicted.cluster.score"
        )]
    res_df['donor'] = sub(".RDS", "", this_query_file)
    donor_id <- sub(".RDS", ".feather", this_query_file)
    write_feather(res_df, paste0(data_path, "human_mtg_query_results/",donor_id))

    end_time <- Sys.time()
    print(paste('Mapping time', end_time - start_time))
    remove(query)
    gc()
}
